### PR TITLE
fold resize of non-splat tosa const iff used once

### PR DESCRIPTION
**Pull request by ljfitz**

### CODE REVIEW

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 1]
```diff
@@ -877,11 +877,23 @@ OpFoldResult ReshapeOp::fold(FoldAdaptor adaptor) {
...
+    return {};
+
+  // Okay to duplicate splat constants.
+  if (operand.isSplat()) {
     return SplatElementsAttr::get(outputTy, operand.getSplatValue<Attribute>());
   }
 
-  return {};
+  // Don't duplicate other constants.
+  if (!(getInput1().hasOneUse() || operand.isSplat()))
```
Comment (mgehre-amd) [inline mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:892]: We already checked isSplat a few lines earlier, so you can remove it here

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 2]
```diff
@@ -877,11 +877,23 @@ OpFoldResult ReshapeOp::fold(FoldAdaptor adaptor) {
     return getInput1();
 
   auto operand = llvm::dyn_cast_if_present<DenseElementsAttr>(adaptor.getInput1());
```
Comment (mgehre-amd) [inline mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:879]: If operand is not a DenseElementsAttr but a DenseResourceElmenetsAttr (aka externalized/elided constant), then you can use
```
+  if (auto input =
+          dyn_cast_if_present<DenseResourceElementsAttr>(adaptor.getInput1())) {
+   return DenseResourceElementsAttr::get(outputTy, input.getRawHandle());
+  }
```
to reshape them.

**Review by mgehre-amd — APPROVED** [review]
Thanks for opening the PR! Just small comments.

**Comment by mgehre-amd** [comment]
I opened an upstream PR with some changes here https://reviews.llvm.org/D154615 to fix the crash I see with make_fx in torch_mlir.

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp (hunk 2 above)
Comment (ljfitz) [reply mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp]: I wonder if in any way there could be a tight coupling between the resource and the dimensions it was created for. I believe in theory there could be. One option would be to specifically support `__elided__`, though AFAIK that has no official support and is just a convenience for developers. 

```
// CHECK-LABEL: @reshape_canonicalize_const_resource_elided
func.func @reshape_canonicalize_const_resource_elided() -> (tensor<3xi32>, tensor<1x3xi32>) {
  %0 = "tosa.const"() {value = dense_resource<__elided__> : tensor<3xi32>} : ()-> tensor<3xi32>
  %1 = "tosa.reshape"(%0) {new_shape = array<i64: 1, 3>} : (tensor<3xi32>) -> tensor<1x3xi32>
  return %0 , %1 : tensor<3xi32>, tensor<1x3xi32>
}

// CHECK-LABEL: @reshape_canonicalize_const_resource_other
func.func @reshape_canonicalize_const_resource_other() -> (tensor<3xi32>, tensor<1x3xi32>) {
  // CHECK: "tosa.reshape"
  %0 = "tosa.const"() {value = dense_resource<__other__> : tensor<3xi32>} : ()-> tensor<3xi32>
  %1 = "tosa.reshape"(%0) {new_shape = array<i64: 1, 3>} : (tensor<3xi32>) -> tensor<1x3xi32>
  return %0 , %1 : tensor<3xi32>, tensor<1x3xi32>
}
```

Lets discuss for a potential follow-up PR.

Comment (mgehre-amd) [reply mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp]: Right, maybe my assumptions are too shaky. I have no further information about dense_resource, so I cannot really judge.

**Pull request merged**